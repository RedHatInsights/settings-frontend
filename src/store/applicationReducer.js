import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { ACTION_TYPES } from '../constants';

export const defaultState = { loaded: false, configLoaded: false, error: null };

export const loading = (store) => {
  return {
    ...store,
    loaded: false,
    schema: [],
  };
};

export const getSchema = (store, { payload }) => {
  return {
    ...store,
    schema: payload,
    loaded: true,
  };
};

export const getConfig = (store, { payload }) => {
  return {
    ...store,
    appsConfig: payload,
    configLoaded: true,
  };
};

export const getSchemaRejected = (store) => ({
  ...store,
  error: true,
});

export default {
  applicationsStore: applyReducerHash(
    {
      [ACTION_TYPES.GET_SCHEMA_FULFILLED]: getSchema,
      [ACTION_TYPES.GET_SCHEMA_PENDING]: loading,
      [ACTION_TYPES.GET_SCHEMA_REJECTED]: getSchemaRejected,
      [ACTION_TYPES.SAVE_VALUES_FULFILLED]: (store) => ({ ...store }),
      [ACTION_TYPES.GET_CONFIG_PENDING]: (store) => ({
        ...store,
        appsConfig: {},
        configLoaded: false,
      }),
      [ACTION_TYPES.GET_CONFIG_FULFILLED]: getConfig,
    },
    defaultState
  ),
};
