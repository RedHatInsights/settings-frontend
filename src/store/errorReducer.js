import { API_ERROR } from '../constants';

const setErrorState = (_, { payload }) => ({
  errorCode: payload,
});

export default {
  [API_ERROR]: setErrorState,
};
