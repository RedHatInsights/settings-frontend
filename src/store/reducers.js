import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import { ACTION_TYPES } from '../constants';

export const defaultState = { loaded: false, configLoaded: false };

export const loading = (store) => {
    return {
        ...store,
        loaded: false,
        schema: []
    };};

export const getSchema = (store, { payload }) => {
    return {
        ...store,
        schema: payload,
        loaded: true
    };
};

export const getConfig = (store, { payload }) => {
    return ({
        ...store,
        appsConfig: payload,
        configLoaded: true
    });
};

export default {
    applicationsStore: applyReducerHash({
        [ACTION_TYPES.GET_SCHEMA_FULFILLED]: getSchema,
        [ACTION_TYPES.GET_SCHEMA_PENDING]: loading,
        [ACTION_TYPES.SAVE_VALUES_FULFILLED]: (store) => ({ ...store }),
        [ACTION_TYPES.GET_CONFIG_PENDING]: (store) => ({ ...store, appsConfig: {}, configLoaded: false }),
        [ACTION_TYPES.GET_CONFIG_FULFILLED]: getConfig
    }, defaultState)
};
