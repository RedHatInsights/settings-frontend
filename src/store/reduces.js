import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import { ACTION_TYPES } from '../constants';

const defaultState = { loaded: false };

const loading = () => {
    return {
        loaded: false,
        schema: {}
    };};

const getSchema = (store, { payload }) => {
    return {
        ...store,
        schema: payload,
        loaded: true
    };};

export default {
    applicationsStore: applyReducerHash({
        [ACTION_TYPES.GET_SCHEMA_FULFILLED]: getSchema,
        [ACTION_TYPES.GET_SCHEMA_PENDING]: loading,
        [ACTION_TYPES.SAVE_VALUES_FULFILLED]: (store) => ({ ...store })
    }, defaultState)
};
