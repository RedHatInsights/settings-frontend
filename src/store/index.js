import ReducerRegistry, {
  applyReducerHash,
} from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import promiseMiddleware from 'redux-promise-middleware';
import notificationsMiddleware from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import errorReducer from './errorReducer';
import logger from 'redux-logger';

const registry = new ReducerRegistry({}, [
  promiseMiddleware,
  notificationsMiddleware(),
  ...(window.insights.chrome.isProd ? [] : [logger]),
]);

registry.register({ errorReducer: applyReducerHash(errorReducer) });

export function getStore() {
  return registry.getStore();
}

export function register(...args) {
  return registry.register(...args);
}

export default registry;
