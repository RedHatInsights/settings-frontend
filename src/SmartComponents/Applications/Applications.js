import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Skeleton, PageHeader, PageHeaderTitle, Main } from '@redhat-cloud-services/frontend-components';
import { Stack, StackItem, Card, CardBody } from '@patternfly/react-core';
import FormRender from '@data-driven-forms/react-form-renderer';
import { formFieldsMapper, layoutMapper } from '@data-driven-forms/pf4-component-mapper';
import registryDecorator from '@redhat-cloud-services/frontend-components-utilities/files/Registry';
import { register } from '../../store';
import reducers  from '../../store/reduces.js';
import { getSchema, saveValues } from '../../actions';

@registryDecorator()
class Applications extends Component {
    constructor(props) {
        super(props);
        register(reducers);
        props.getSchema(props.match.params.id);
        this.state = {
            appName: this.props.match.params.id
        };
    }

    renderForms(schemas, appName, loaded) {
        return (
            <Stack>
                { loaded
                    ?  schemas.map((schema, i) => (
                        <StackItem key={ `settings-form-${i}` }>
                            <Card>
                                <CardBody>
                                    <FormRender
                                        formFieldsMapper={ formFieldsMapper }
                                        layoutMapper={ layoutMapper }
                                        schema={ schema }
                                        onSubmit={ (values) => this.props.saveValues(appName, values) }
                                        canReset
                                    />
                                </CardBody>
                            </Card>
                        </StackItem>
                    ))
                    : <StackItem>
                        <Card>
                            <CardBody>
                                <Skeleton size='lg' />
                            </CardBody>
                        </Card>
                    </StackItem>
                }
            </Stack>);
    }

    render() {
        const { appName }  = this.state;
        const { loaded, schema } = this.props;
        return (
            <React.Fragment>
                <PageHeader>
                    <PageHeaderTitle title='Applications Settings'/>
                    <p>{ `Settings for ${ appName }` }</p>
                </PageHeader>
                <Main>
                    { this.renderForms(schema, appName, loaded) }
                </Main>
            </React.Fragment>
        );
    }
}

Applications.propTypes = {
    match: PropTypes.shape({
        params: PropTypes.shape({
            id: PropTypes.string
        })
    }),
    getSchema: PropTypes.func,
    loaded: PropTypes.bool,
    saveValues: PropTypes.func,
    schema: PropTypes.arrayOf(PropTypes.object)
};

function mapStateToProps({ applicationsStore }) {
    return {
        schema: applicationsStore && applicationsStore.schema,
        loaded: applicationsStore && applicationsStore.loaded
    };
}

function mapDispatchToProps(dispatch) {
    return {
        getSchema: (application) => dispatch(getSchema(application)),
        saveValues: (application, values) => dispatch(saveValues(application, values))
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Applications);
