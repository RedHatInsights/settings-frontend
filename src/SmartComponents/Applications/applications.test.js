import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import { componentTypes, validatorTypes } from '@data-driven-forms/react-form-renderer';
import Applications from './Applications';
import { init } from '../../store';

const emptyState = {
    applicationsStore: {
        loaded: false,
        configLoaded: false,
        schema: [],
        appsConfig: {}
    }
};
const mockState = {
    applicationsStore: {
        loaded: true,
        schema: [{
            fields: [{
                name: 'email',
                label: 'Email',
                component: componentTypes.TEXT_FIELD,
                isRequired: true,
                validate: [{
                    type: validatorTypes.REQUIRED
                }],
                initialValue: ''
            }, {
                name: 'hideSatelliteSystems',
                label: 'Hide Satellite systems',
                component: componentTypes.SWITCH,
                type: 'checkbox',
                initialValue: false
            }]
        }],
        configLoaded: true,
        appsConfig: {
            testapp: {
                frontend: {
                    title: 'Test app'
                }
            }
        }
    }
};
let mockStore;
init();

describe('Applications', () => {
    beforeEach(() => {
        mockStore = configureStore([ promiseMiddleware(), notificationsMiddleware() ]);
    });

    it('Render applications with no data', () => {
        const store = mockStore({});
        const wrapper = render(<Provider store={ store }>
            <Applications match={ { params: { id: 'testapp' }} }/>
        </Provider>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Render applications with emptyState', () => {
        const store = mockStore(emptyState);
        const wrapper = render(<Provider store={ store }>
            <Applications match={ { params: { id: 'testapp' }} } />
        </Provider>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Render applications with mockState', () => {
        const store = mockStore(mockState);
        const wrapper = render(<Provider store={ store }>
            <Applications match={ { params: { id: 'testapp' }} } />
        </Provider>);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should emit type-success notification on saving a form', () => {
        const store = mockStore(mockState);
        const wrapper = mount(
            <Provider store={ store }>
                <Applications appsConfig={ {} } match={ { params: { id: 'testapp' }} } />
            </Provider>,
        );
        const input = wrapper.find('input#email');
        input.getDOMNode().value = 'value';
        input.simulate('change');
        wrapper.update();
        wrapper.find('.pf-m-primary').simulate('click');
        const expectedPayload = [
            expect.anything(),
            expect.objectContaining({
                meta: { notifications: {
                    fulfilled: {
                        description: 'Settings for Red Hat Insights were replaced with new values.',
                        dismissable: true,
                        title: 'Application settings saved',
                        variant: 'success'
                    }
                }}
            })
        ];
        expect(store.getActions()).toEqual(expectedPayload);
    });
});
