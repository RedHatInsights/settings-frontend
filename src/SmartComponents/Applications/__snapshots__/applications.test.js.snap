// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applications Render applications with emptyState 1`] = `
Array [
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      widget-type="InsightsPageHeaderTitle"
    >
       Applications settings 
    </h1>
    <div
      class="ins-c-skeleton ins-c-skeleton__sm"
    >
       
    </div>
  </section>,
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <article
          class="pf-c-card"
        >
          <div
            class="pf-c-card__body"
          >
            <div
              class="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </div>
        </article>
      </div>
    </div>
  </section>,
]
`;

exports[`Applications Render applications with mockState 1`] = `
Array [
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      widget-type="InsightsPageHeaderTitle"
    >
       Applications settings 
    </h1>
    <p>
      Settings for Test app
    </p>
  </section>,
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <article
          class="pf-c-card"
        >
          <div
            class="pf-c-card__body"
          >
            <form
              class="pf-c-form"
              novalidate=""
            >
              <div
                class="pf-c-form__group"
              >
                <label
                  class="pf-c-form__label"
                  for="email"
                >
                  <span
                    class="pf-c-form__label-text"
                  >
                    Email
                  </span>
                  <span
                    aria-hidden="true"
                    class="pf-c-form__label-required"
                  >
                    *
                  </span>
                </label>
                <input
                  aria-invalid="false"
                  class="pf-c-form-control"
                  id="email"
                  label="Email"
                  name="email"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="pf-c-form__group"
              >
                <label
                  class="pf-c-switch"
                  for="hideSatelliteSystems"
                >
                  <input
                    aria-label=""
                    aria-labelledby="hideSatelliteSystems-on"
                    class="pf-c-switch__input"
                    id="hideSatelliteSystems"
                    name="hideSatelliteSystems"
                    type="checkbox"
                    value="false"
                  />
                  <span
                    class="pf-c-switch__toggle"
                  />
                  <span
                    aria-hidden="true"
                    class="pf-c-switch__label pf-m-on"
                    id="hideSatelliteSystems-on"
                  >
                    Hide Satellite systems
                  </span>
                  <span
                    aria-hidden="true"
                    class="pf-c-switch__label pf-m-off"
                    id="hideSatelliteSystems-off"
                  >
                    Hide Satellite systems
                  </span>
                </label>
              </div>
              <div
                class="pf-c-form__group pf-m-action"
              >
                <div
                  class="pf-c-form__actions"
                >
                  <button
                    class="pf-c-button pf-m-primary"
                    type="submit"
                  >
                    Submit
                  </button>
                  <button
                    class="pf-c-button pf-m-secondary"
                    disabled=""
                    type="button"
                  >
                    Reset
                  </button>
                </div>
              </div>
            </form>
          </div>
        </article>
      </div>
    </div>
  </section>,
]
`;

exports[`Applications Render applications with no data 1`] = `
Array [
  <section
    class="pf-l-page-header pf-c-page-header pf-l-page__main-section pf-c-page__main-section pf-m-light"
    widget-type="InsightsPageHeader"
  >
    <h1
      class="pf-c-title pf-m-2xl"
      widget-type="InsightsPageHeaderTitle"
    >
       Applications settings 
    </h1>
    <div
      class="ins-c-skeleton ins-c-skeleton__sm"
    >
       
    </div>
  </section>,
  <section
    class="pf-l-page__main-section pf-c-page__main-section"
    page-type=""
  >
    <div
      class="pf-l-stack pf-m-gutter"
    >
      <div
        class="pf-l-stack__item"
      >
        <article
          class="pf-c-card"
        >
          <div
            class="pf-c-card__body"
          >
            <div
              class="ins-c-skeleton ins-c-skeleton__lg"
            >
               
            </div>
          </div>
        </article>
      </div>
    </div>
  </section>,
]
`;
